File: tomfoolery/Notepad_Clone/notepad_clone.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <algorithm> // Needed for std::remove

// notepad clone? 
// CRUD notes? 
// organize by categories/ tags? 
// search real time 

namespace Notepad_Clone {
    class Tag {
    public:
        std::vector<std::string> tags;

        void AddTag() {
            std::string new_tag;
            std::cout << "Enter a tag: ";
            std::cin >> new_tag; 
            tags.push_back(new_tag);
        }

        void ReadTag() const {
            for (const auto& tag : tags) {
                std::cout << tag << std::endl;
            }
        }

        void DeleteTag() {
            std::string user_input;
            std::cout << "What tag to delete: ";
            std::cin >> user_input;
            auto tag_remove = std::remove(tags.begin(), tags.end(), user_input);
            if (tag_remove != tags.end()) {
                tags.erase(tag_remove, tags.end());
            } else {
                std::cout << "Tag not found." << std::endl;
            }
        }
    };

    class Note {
    public:
        std::string title;
        std::string content;
        Tag tags;

        void SetTitle() {
            std::cout << "Enter a new title: ";
            std::getline(std::cin, title);
        }

        void SetContent() {
            std::cout << "Enter content for the note: ";
            std::getline(std::cin, content);
        }

        void ShowDetails() const {
            std::cout << "Title: " << title << "\nContent: " << content << std::endl;
            std::cout << "Tags: " << std::endl;
            tags.ReadTag();
        }

        void AddTag() {
            tags.AddTag(); // Delegate to Tag class
        }
    };

    class Notepad {
        std::map<std::string, Note> notes;

    public:
        void AddNote() {
            Note new_note;
            new_note.SetTitle();
            new_note.SetContent();

            notes[new_note.title] = new_note;

            std::cout << "Note added: " << new_note.title << std::endl;
        }

        Note* GetNote(const std::string &title) {
            auto it = notes.find(title);
            if (it != notes.end()) { // Corrected: use 'end()'
                return &(it->second); // Return pointer to the Note object
            }
            return nullptr; // Return null if note not found
        }

        void ShowNotes() const {
            for (const auto& pair : notes) {
                std::cout << "Note title: " << pair.first << std::endl; // Corrected: use 'std::endl'
            }
        }
    };
}

int main() {
    Notepad_Clone::Notepad notepad; // Create Notepad object

    // Example usage
    notepad.AddNote(); // User sets title and content
    Notepad_Clone::Note* note = notepad.GetNote("Your Note Title"); // Replace with actual title

    if (note) {
        note->AddTag(); // Add a tag
        note->ShowDetails(); // Show title, content, and tags
    }

    notepad.ShowNotes(); // Show all notes
    return 0;
}

File: tomfoolery/Notepad_Clone/NotepadClone.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

class Tag {
    private List<String> tags;

    public Tag() {
        tags = new ArrayList<>();
    }

    public void addTag(Scanner scanner) {
        System.out.print("Enter a tag: ");
        String newTag = scanner.nextLine();
        tags.add(newTag);
    }

    public void readTags() {
        for (String tag : tags) {
            System.out.println(tag);
        }
    }

    public void deleteTag(Scanner scanner) {
        System.out.print("What tag to delete: ");
        String userInput = scanner.nextLine();
        if (tags.remove(userInput)) {
            System.out.println("Tag deleted.");
        } else {
            System.out.println("Tag not found.");
        }
    }
}

class Note {
    private String title;
    private String content;
    private Tag tags;

    public Note() {
        tags = new Tag();
    }

    public void setTitle(Scanner scanner) {
        System.out.print("Enter a new title: ");
        title = scanner.nextLine();
    }

    public void setContent(Scanner scanner) {
        System.out.print("Enter content for the note: ");
        content = scanner.nextLine();
    }

    public void showDetails() {
        System.out.println("Title: " + title);
        System.out.println("Content: " + content);
        System.out.println("Tags: ");
        tags.readTags();
    }

    public String getTitle() {
        return title;
    }

    public Tag getTags() {
        return tags;
    }
}

class Notepad {
    private Map<String, Note> notes;

    public Notepad() {
        notes = new HashMap<>();
    }

    public void addNote(Scanner scanner) {
        Note newNote = new Note();
        newNote.setTitle(scanner);
        newNote.setContent(scanner);
        notes.put(newNote.getTitle(), newNote);
        System.out.println("Note added: " + newNote.getTitle());
    }

    public Note getNote(String title) {
        return notes.get(title);
    }

    public void showNotes() {
        for (String title : notes.keySet()) {
            System.out.println("Note title: " + title);
        }
    }
}

public class NotepadClone {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Notepad notepad = new Notepad();

        // Example usage
        notepad.addNote(scanner);
        System.out.print("Enter the title of the note you'd like to access: ");
        String title = scanner.nextLine();
        Note note = notepad.getNote(title);

        if (note != null) {
            note.getTags().addTag(scanner);  // Add a tag
            note.showDetails();               // Show title, content, and tags
        } else {
            System.out.println("Note not found.");
        }

        notepad.showNotes();  // Show all notes
        scanner.close();
    }
}
